### ISSUE TITLE
[BUG] Existing files recognized but still queued

### ISSUE THREAD

**Issue by jonathanm-ur** [Tzahi12345/YoutubeDL-Material#480]
**Describe the bug**
After checking a subscription a subset of files is detected as new, but files are already downloaded, resulting in an error.

Log showing behaviour

> 2021-10-07T16:31:34.301Z VERBOSE: Subscription: finished check for Frog Leap Studios
> 2021-10-07T16:31:38.912Z INFO: Skipping adding file subscriptions/channels/Frog Leap Studios/NEW ALBUM IS OUT!.mp4 for subscription Frog Leap Studios as a file with that path already exists.
> 2021-10-07T16:31:44.075Z INFO: Skipping adding file subscriptions/channels/Frog Leap Studios/NEW ALBUM & GIVE AWAY!.mp4 for subscription Frog Leap Studios as a file with that path already exists.
> 2021-10-07T16:31:45.607Z INFO: Skipping adding file subscriptions/channels/Frog Leap Studios/What´s In The Box.mp4 for subscription Frog Leap Studios as a file with that path already exists.
> 2021-10-07T16:31:48.085Z INFO: Skipping adding file subscriptions/channels/Frog Leap Studios/ALBUM RELEASE!.mp4 for subscription Frog Leap Studios as a file with that path already exists.
> 2021-10-07T16:31:54.686Z DEBUG: Inserted doc into download_queue
> 2021-10-07T16:31:54.688Z DEBUG: Inserted doc into download_queue
> 2021-10-07T16:31:54.690Z DEBUG: Inserted doc into download_queue
> 2021-10-07T16:31:54.692Z DEBUG: Inserted doc into download_queue
> TypeError: Cannot read property 'forEach' of undefined
> at setFreshUploads (/app/subscriptions.js:485:16)
> at /app/subscriptions.js:306:27
> at runMicrotasks (<anonymous>)
> at processTicksAndRejections (internal/process/task_queues.js:95:5)
> 2021-10-07T16:31:55.094Z VERBOSE: Collecting info for download e5a10241-047a-4e80-ab89-9f95b5ac19ed
> 2021-10-07T16:31:55.094Z VERBOSE: Collecting info for download e5a10241-047a-4e80-ab89-9f95b5ac19ed
> 2021-10-07T16:31:55.868Z VERBOSE: youtube-dl args being used: -o,subscriptions/channels/Frog Leap Studios/%(title)s.%(ext)s,--write-info-json,--print-json,-f,bestvideo+bestaudio,--merge-output-format,mp4,--download-archive,/app/subscriptions/archives/Frog Leap Studios/merged_video.txt,--write-thumbnail,--no-clean-infojson
> 2021-10-07T16:32:07.105Z VERBOSE: Downloading e5a10241-047a-4e80-ab89-9f95b5ac19ed
> 2021-10-07T16:32:07.105Z VERBOSE: Downloading e5a10241-047a-4e80-ab89-9f95b5ac19ed
> 2021-10-07T16:32:11.058Z DEBUG: Video download delay: 3.95 seconds.
> 2021-10-07T16:32:11.060Z WARN: No output received for video download, check if it exists in your archive.
> 2021-10-07T16:32:11.106Z VERBOSE: Collecting info for download cdf04f48-08d4-483e-8803-abe59e4e0126
> 2021-10-07T16:32:11.111Z VERBOSE: youtube-dl args being used: -o,subscriptions/channels/Frog Leap Studios/%(title)s.%(ext)s,--write-info-json,--print-json,-f,bestvideo+bestaudio,--merge-output-format,mp4,--download-archive,/app/subscriptions/archives/Frog Leap Studios/merged_video.txt,--write-thumbnail,--no-clean-infojson
> 2021-10-07T16:32:23.107Z VERBOSE: Downloading cdf04f48-08d4-483e-8803-abe59e4e0126
> 2021-10-07T16:32:27.068Z DEBUG: Video download delay: 3.959 seconds.
> 2021-10-07T16:32:27.070Z WARN: No output received for video download, check if it exists in your archive.
> 2021-10-07T16:32:27.105Z VERBOSE: Collecting info for download b2ebf4a7-3044-4356-8d91-2033be2a003d
> 2021-10-07T16:32:27.110Z VERBOSE: youtube-dl args being used: -o,subscriptions/channels/Frog Leap Studios/%(title)s.%(ext)s,--write-info-json,--print-json,-f,bestvideo+bestaudio,--merge-output-format,mp4,--download-archive,/app/subscriptions/archives/Frog Leap Studios/merged_video.txt,--write-thumbnail,--no-clean-infojson
> 2021-10-07T16:32:38.107Z VERBOSE: Downloading b2ebf4a7-3044-4356-8d91-2033be2a003d
> 2021-10-07T16:32:42.407Z DEBUG: Video download delay: 4.298 seconds.
> 2021-10-07T16:32:42.409Z WARN: No output received for video download, check if it exists in your archive.
> 2021-10-07T16:32:43.108Z VERBOSE: Collecting info for download 7d61a009-a1fb-4973-aa39-ab7e6e7df636
> 2021-10-07T16:32:43.112Z VERBOSE: youtube-dl args being used: -o,subscriptions/channels/Frog Leap Studios/%(title)s.%(ext)s,--write-info-json,--print-json,-f,bestvideo+bestaudio,--merge-output-format,mp4,--download-archive,/app/subscriptions/archives/Frog Leap Studios/merged_video.txt,--write-thumbnail,--no-clean-infojson
> 2021-10-07T16:32:43.108Z VERBOSE: Collecting info for download 7d61a009-a1fb-4973-aa39-ab7e6e7df636
> 2021-10-07T16:32:43.112Z VERBOSE: youtube-dl args being used: -o,subscriptions/channels/Frog Leap Studios/%(title)s.%(ext)s,--write-info-json,--print-json,-f,bestvideo+bestaudio,--merge-output-format,mp4,--download-archive,/app/subscriptions/archives/Frog Leap Studios/merged_video.txt,--write-thumbnail,--no-clean-infojson
> 2021-10-07T16:32:56.113Z VERBOSE: Downloading 7d61a009-a1fb-4973-aa39-ab7e6e7df636
> 2021-10-07T16:32:56.113Z VERBOSE: Downloading 7d61a009-a1fb-4973-aa39-ab7e6e7df636
> 2021-10-07T16:33:00.021Z DEBUG: Video download delay: 3.906 seconds.
> 2021-10-07T16:33:00.023Z WARN: No output received for video download, check if it exists in your archive.
> 

Screenshot of download manager.
![image](https://user-images.githubusercontent.com/64561015/136427608-670aad03-b182-443f-980c-d1646383c82f.png)

**Environment**
- YoutubeDL-Material version
nightly as of this post

This behaviour is repeated across many of my subs, I just pulled this out as a good representative example.

**Comment by jonathanm-ur** [comment]
It also makes the new downloads page almost useless, even at 20 items per page I'm only getting 1 good download per page on average.
![image](https://user-images.githubusercontent.com/64561015/136955455-d05d750a-6d5f-4ced-a5b1-fd44160930bd.png)

Lots of noise, no signal.

Oh, while we're here, the downloads page action buttons could use some love, the current "Clear finished downloads" button isn't enough specificity.

Ideally, there would be buttons for clearing only good downloads, and requeueing all failed downloads. This assumes that the current repeat offenders caught by archive wouldn't show up on the list any more, only legitimate fails.

**Comment by Tzahi12345** [comment]
Apologies for the delay @jonathanm-ur. Perhaps we can do a dialog for clearing downloads?

Not sure how to deal with the case of lots of downloads failed due to the archive. "Redownload failed downloads" could work, but in this case you'd want to be more specific no?

**Comment by GlassedSilver** [comment]
Failed downloads due to archive is the wrong way to look at this.

The downloads are successfully skipped due to a lack of necessity. ;)

**Comment by jonathanm-ur** [comment]
I'd be fine if they were only logged instead of clogging up the download list, as I could more easily ignore them.

Could they possibly be not listed as failed in the downloads, since apparently they didn't fail in the past at some point?

Genuine failures should be available to retry, as it is right now the download list is full of "failures" that are really just previous successes that are "skipped due to lack of necessity."

Also, I still have a concern about what exactly is causing the redownload attempt, since it happens to a handful of videos quite consistently. I'm guessing the archive feature isn't catching them for some reason?

All I'm saying is, if the cause is out of your control, can we handle it better in the app somehow?

**Comment by GlassedSilver** [comment]
> Also, I still have a concern about what exactly is causing the redownload attempt, since it happens to a handful of videos quite consistently. I'm guessing the archive feature isn't catching them for some reason?

Yeah I get this myself on a few videos. I have a sneaky suspicion what might cause this, but I‘m not sure.

I‘ll do some investigation on my instance later tonight. 👍

**Comment by GlassedSilver** [comment]
I think I got it.

This is happening in subscriptions, the archive that we save to these days is subscriptions/archives/archive_video.txt as well as blacklist_video.txt and merged_video.txt respectively in the same folder per subscription.

Before we also had (and maybe this is a switching from ytdl to yt-dlp thing? idk, you tell us @Tzahi12345) archive.txt, but that file never changed since Sept. 2021.

I sus for the queuing we look at archive.txt and yt-dlp then successfully checks the others for DL conflicts.

**Comment by jonathanm-ur** [comment]
Still happening on my instance FWIW.

Do I need to manipulate any of the archive txt lists to get this fix to work for me?

**Comment by GlassedSilver** [comment]
> Still happening on my instance FWIW.
> 
> Do I need to manipulate any of the archive txt lists to get this fix to work for me?

It's very possible that now archive.txt not getting looked at anymore but the archive file name we now use to save to ever since September last year means that some older videos that were in archive.txt but not in the new ones get re-downloaded.

What you can do is merge IDs from archive.txt into the respective new files per subscription. I also intend to make a fix-script available for this and/or we integrate a task on the Web UI. :)

**Comment by Tzahi12345** [comment]
I made some docs for archiving functionality as even I was confused by the whole system. In short it needs to be reworked but I made a table for where the relevant paths are:

https://github.com/Tzahi12345/YoutubeDL-Material/wiki/Archive

There is an issue where sometimes subscriptions are assumed to be using the `archive.txt` file. Working on a fix for this.

**Comment by Tzahi12345** [comment]
@jonathanm-ur Added a fix in #657 for the error you saw in your logs:

```
TypeError: Cannot read property 'forEach' of undefined
at setFreshUploads (/app/subscriptions.js:485:16)
at /app/subscriptions.js:306:27
at runMicrotasks ()
at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

Commit: https://github.com/Tzahi12345/YoutubeDL-Material/pull/657/commits/a4a004547531617bc797584068f82020d41690c0

Will be merged shortly with v4.3 release!
